Issue #13398 - Improve LoginAuthenticators to handle Proxy-Authenticate

znight1020 opened the pull request:
This pull request resolves issue #13398 by enhancing server-side authenticators to support proxy authentication.

### Summary of Changes

1.  **Authenticator Enhancements:**
    *   Server-side authenticators like `BasicAuthenticator`, `DigestAuthenticator` and `SPNEGOAuthenticator ` no longer handle only hardcoded `WWW-Authenticate` headers.
    *   A new `setProxyMode(boolean)` method has been added to the base `LoginAuthenticator` in both `jetty-security` and `jetty-ee9-security`.
    *   When proxy mode is enabled, these authenticators now dynamically issue `407` challenges with `Proxy-Authenticate` headers and process credentials from the `Proxy-Authorization` header, using new internal helper methods.

2.  **Test Refactoring:**
    *   The parent `HttpClientAuthenticationTest` has been refactored to use a new `protected boolean isProxyMode()` method (which defaults to `false`), allowing it to test both authentication modes.
    *   A new `HttpClientProxyAuthenticationTest` subclass is introduced, which simply overrides `isProxyMode()` to return `true`, reusing the entire parent test suite for comprehensive validation of the new proxy mode.

Resolves #13398

znight1020 committed:
Improve LoginAuthenticators to handle Proxy-Authenticate

znight1020 committed:
Add tests for Proxy-Authenticate mode

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java:
```diff
@@ -166,11 +176,11 @@ public void onSuccess(Request request)
         client.getRequestListeners().addListener(requestListener);
 
-        // Request without Authentication causes a 401
+        // Request without Authentication causes 401 / 407
```
```suggestion
        // Request without authentication causes 401 / 407
```

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java:
```diff
@@ -70,12 +70,22 @@ public void startBasic(Scenario scenario, Handler handler, Charset charset) thro
     public void startDigest(Scenario scenario, Handler handler) throws Exception
     {
-        start(scenario, new DigestAuthenticator(), handler);
+        DigestAuthenticator authenticator = new DigestAuthenticator();
```
Please store the `authenticator` as a field of type `LoginAuthenticator`.
In this way, in the tests you can call `authenticator.getUnauthorizedStatusCode()`, `authenticator.getAuthorizationHeader()`, etc. without duplicating it in the test class.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java:
```diff
@@ -810,6 +813,57 @@ public void testSingleChallengeLooksLikeMultipleChallenges()
         assertEquals(headerInfo.getParameter("nonce"), "1523430383=");
     }
 
+    private int getUnauthorizedStatus()
```
Remove this method in favor of calling `authenticator.getUnauthorizedStatusCode()`.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java:
```diff
@@ -430,7 +441,11 @@ public void testPreemptedAuthentication(Scenario scenario) throws Exception
+        authenticationStore.addAuthenticationResult(
+            isProxyMode()
+                ? new BasicAuthentication.BasicResult(uri, HttpHeader.PROXY_AUTHORIZATION, "basic", "basic")
+                : new BasicAuthentication.BasicResult(uri, "basic", "basic")
```
Replace this by calling `authenticator.getAuthorizationHeader()`.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java:
```diff
@@ -589,8 +587,13 @@ public boolean handle(org.eclipse.jetty.server.Request request, org.eclipse.jett
-                response.getHeaders().put(HttpHeader.WWW_AUTHENTICATE, authType);
+                response.setStatus(getUnauthorizedStatus());
+                response.getHeaders().put(
+                    isProxyMode()
+                        ? HttpHeader.PROXY_AUTHENTICATE
+                        : HttpHeader.WWW_AUTHENTICATE,
```
Replace this by calling `authenticator.getChallengeHeader()`.

znight1020 replied:
I've applied it throughout the other test methods.

However, I'm facing a small issue specifically in the `testInfiniteAuthentication` method. Unlike the other authentication tests that call `startBasic()` or `startDigest()` (which initialize the `this.authenticator` field), this particular test calls the more fundamental `start(scenario, handler)` method directly.

As a result, the `this.authenticator` field remains `null` within the scope of this test, as shown below:

```java
// Initialize Authenticator
private LoginAuthenticator authenticator;

public void startBasic(Scenario scenario, Handler handler, Charset charset) throws Exception
{
    BasicAuthenticator basicAuthenticator = new BasicAuthenticator();
    ...
    authenticator = basicAuthenticator;

    start(scenario, authenticator, handler);
}
```
```java
// In testInfiniteAuthentication:
start(scenario, new Handler.Abstract() 
{
    @Override
    public boolean handle(...) 
    {
        // `authenticator` field is null.
        response.getHeaders().put(authenticator.getChallengeHeader(), authType); // This would cause an NPE.
        ...
    }
});
...
assertEquals(authenticator.getUnauthorizedStatusCode(), response.getStatus()); // This would cause an NPE.
```

Given this, would it be acceptable to use `isProxyMode()` directly in the handler for this specific test, or I change the test to use `startBasic()` so that the `authenticator` field is initialized?

sbordet replied:
Please use `isProxyMode()` for this test. You would need a ternary expression in the `Handler`, and in the last assertion.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/LoginAuthenticator.java:
```diff
@@ -39,11 +41,43 @@ public abstract class LoginAuthenticator implements Authenticator
     {
     }
 
+    public boolean isProxy()
```
Rename to `isProxyMode()`.
Also please add minimal javadocs and `@see #setProxyMode(boolean)`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/LoginAuthenticator.java:
```diff
@@ -39,11 +41,43 @@ public abstract class LoginAuthenticator implements Authenticator
+        return _proxy;
+    }
+
+    public void setProxy(boolean proxy)
```
Rename to `setProxyMode()`, and add javadocs explaining the effects of this boolean on the other methods: `getAuthorizationHeader()`, `getChallengeHeader()` and `getUnauthorizedStatusCode()`.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/LoginAuthenticator.java:
```diff
@@ -39,11 +41,43 @@ public abstract class LoginAuthenticator implements Authenticator
+        _proxy = proxy;
+    }
+
+    protected final HttpHeader getAuthorizationHeader()
```
Add javadocs.

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/LoginAuthenticator.java:
```diff
@@ -39,11 +41,43 @@ public abstract class LoginAuthenticator implements Authenticator
+            : HttpHeader.WWW_AUTHENTICATE;
+    }
+
+    protected final int getChallengeStatusCode()
```
Rename to `getUnauthorizedStatusCode()`.

sbordet commented on jetty-ee9/jetty-ee9-security/src/main/java/org/eclipse/jetty/ee9/security/authentication/LoginAuthenticator.java:
```diff
@@ -40,11 +42,43 @@ public abstract class LoginAuthenticator implements Authenticator
     {
     }
 
+    public boolean isProxy()
```
Same remarks as the core `LoginAuthenticator` with respect to renamings.

sbordet reviewed (changes requested):
@znight1020 excellent work!
Please follow the few renamings that I suggested, and then we should be good to merge.

znight1020 commented:
Thanks for the feedback, @sbordet!

I've implemented the requested renamings and Javadoc improvements, but before I push the changes, I had a quick question regarding `testInfiniteAuthentication`, which I've detailed in the unresolved conversation above.

sbordet commented:
@znight1020 I don't see the changes after review.
Did you push them?

znight1020 committed:
Updates from review

znight1020 commented:
Sorry for that, @sbordet. I've just pushed the requested changes now!

To get the tests passing, I've temporarily updated `testInfiniteAuthentication()` to use `startBasic()`, which resolves the `NullPointerException`. However, my original question from the thread still stands: https://github.com/jetty/jetty.project/pull/13452#discussion_r2272831659

Could you please take a look at the reasoning in that thread when you have a moment?

sbordet commented on jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/LoginAuthenticator.java:
```diff
@@ -39,11 +41,77 @@ public abstract class LoginAuthenticator implements Authenticator
+     * <ul>
+     *  <li>{@link #getChallengeHeader()} will return {@code Proxy-Authenticate}.</li>
+     *  <li>{@link #getUnauthorizedStatusCode()} will return {@code 407}.</li>
+     *  <li>{@link #getAuthorizationHeader()} will read the {@code Proxy-Authorization} header.</li>
```
```suggestion
     *  <li>{@link #getAuthorizationHeader()} will return the {@code Proxy-Authorization} header.</li>
```

sbordet commented on jetty-ee9/jetty-ee9-security/src/main/java/org/eclipse/jetty/ee9/security/authentication/LoginAuthenticator.java:
```diff
@@ -40,11 +42,77 @@ public abstract class LoginAuthenticator implements Authenticator
+     * <ul>
+     *  <li>{@link #getChallengeHeader()} will return {@code Proxy-Authenticate}.</li>
+     *  <li>{@link #getUnauthorizedStatusCode()} will return {@code 407}.</li>
+     *  <li>{@link #getAuthorizationHeader()} will read the {@code Proxy-Authorization} header.</li>
```
```suggestion
     *  <li>{@link #getAuthorizationHeader()} will return the {@code Proxy-Authorization} header.</li>
```

sbordet reviewed (changes requested):
Just a typo in the javadocs, and fix `testInfiniteAuthentication()` and we should be good.

znight1020 committed:
Updates from review

znight1020 commented:
I've corrected the typo in the Javadocs and updated `testInfiniteAuthentication`.
Thanks again for all the feedback and guidance through the process!

sbordet commented:
@znight1020 excellent work!
This PR is approved, we'll wait on the CI build and then merge it.
Thanks!
